runtime/common/src/lib.rs

```diff
@@ -58,3 +64,62 @@ parameter_types! {
 		.saturating_sub(Perbill::from_percent(10)) * MaximumBlockWeight::get();
 	pub const MaximumBlockLength: u32 = 5 * 1024 * 1024;
 }
+
+
+/// Apply a given filter to transactions.
+pub struct TransactionCallFilter<T: Filter<Call>, Call>(PhantomData<(T, Call)>);
+
+impl<F: Filter<Call>, Call> Default for TransactionCallFilter<F, Call> {
+	fn default() -> Self { Self::new() }
+}
+impl<F: Filter<Call>, Call> Encode for TransactionCallFilter<F, Call> {
+	fn using_encoded<R, FO: FnOnce(&[u8]) -> R>(&self, f: FO) -> R { f(&b""[..]) }
+}
+impl<F: Filter<Call>, Call> Decode for TransactionCallFilter<F, Call> {
+	fn decode<I: codec::Input>(_: &mut I) -> Result<Self, codec::Error> { Ok(Self::new()) }
+}
+impl<F: Filter<Call>, Call> Clone for TransactionCallFilter<F, Call> {
+	fn clone(&self) -> Self { Self::new() }
+}
+impl<F: Filter<Call>, Call> Eq for TransactionCallFilter<F, Call> {}
+impl<F: Filter<Call>, Call> PartialEq for TransactionCallFilter<F, Call> {
+	fn eq(&self, _: &Self) -> bool { true }
+}
+impl<F: Filter<Call>, Call> sp_std::fmt::Debug for TransactionCallFilter<F, Call> {
+	fn fmt(&self, _: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result { Ok(()) }
+}
+
+impl<F: Filter<Call> + Send + Sync, Call: Dispatchable + Send + Sync>
+	SignedExtension for TransactionCallFilter<F, Call>
+{
+	const IDENTIFIER: &'static str = "OnlyStakingAndClaims";
+	type AccountId = AccountId;
+	type Call = Call;
+	type AdditionalSigned = ();
+	type Pre = ();
+
+	fn additional_signed(&self) -> sp_std::result::Result<(), TransactionValidityError> { Ok(()) }
+
+	fn validate(
+		&self, _:
+		&Self::AccountId,
+		call: &Call,
+		_: &DispatchInfoOf<Self::Call>,
+		_: usize
+	)
+		-> TransactionValidity
+	{
+		if F::filter(call) {
+			Ok(Default::default())
+		} else {
+			Err(InvalidTransaction::Custom(ValidityError::NoPermission.into()).into())
+		}
+	}
```

Comment:
Shall we filter unsigned transactions as well? I think currently we don't have that many and the ones we have are whitelisted anyway, but might be more future proof to add it here as well.